### ISSUE TITLE
Parametrise task results per environment

### ISSUE THREAD

**Issue by n-batalha** [PrefectHQ/prefect#2714]
TL;DR: we cannot parametrise s3 bucket results per task per environment (prod/staging/etc).

Note: this is a continuation from [Slack](https://prefect-community.slack.com/archives/CL09KU1K7/p1591198856334400) where I was advised by @cicdw that you are working on a new deployment protocol that would enable this, and this ticket was opened to track this

---

We need to deploy Prefect in various environments (prod, staging, dev etc). We will try to isolate these entirely or as much as possible, which means that s3 buckets, dask clusters etc will all vary per environment. Taking the s3 buckets example,we want to add persistency and caching to tasks, so we do

```
@task(result=S3Result(bucket=<REPLACE_ME>)
```

To define the bucket per environment we tried, on `config.toml` (since buckets will be reused across many tasks, as will the dask clusters etc):

```
transient_data_bucket = "${environments.${environment}.transient_data_bucket}"

[environments]

    [environments.dev-nelson]
        transient_data_bucket = "REDACTED"

    [environments.prod]
        transient_data_bucket = "REDACTED"

```

where `environment` is another variable provided via `-e PREFECT__ENVIRONMENT=dev-nelson` to a local agent. Running everything locally, we can obtain `prefect.config.transient_data_bucket` inside a task, but when defining the bucket as `S3Result(bucket=prefect.config.transient_data_bucket)` `boto` fails with `Invalid bucket name ""`.
